### PR TITLE
API for various compression codecs

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
First draft of API.
@trustin and @normanmaurer please review it.

### CODE REVIEW

**Comment by idelpivnitskiy** [comment]
@normanmaurer @trustin 
I see that you prepare 4.1.0.CR1, great news!
Could we complete this work?

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * Creates a new {@link CompressionEncoder} or a new {@link CompressionDecoder}
+ * with specified {@link CompressionFormat}.
+ */
+public final class CompressionCodecFactory {
+
+    public static CompressionEncoder newEncoder(CompressionFormat format) {
+        if (format == null) {
+            throw new NullPointerException("format");
+        }
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java]: ObjectUtil.checkNonNull(...)

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                return new LzmaFrameEncoder();
+            default:
+                throw new IllegalArgumentException("Format " + format + " is not supported");
+        }
+    }
+
+    public static CompressionDecoder newDecoder(CompressionFormat format) {
+        if (format == null) {
+            throw new NullPointerException("format");
+        }
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java]: ObjectUtil.checkNonNull(...)

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2014 The Netty Project
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java]: 2015

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            case FASTLZ:
+                return new FastLzFrameEncoder();
+            case LZMA:
+                return new LzmaFrameEncoder();
+            default:
+                throw new IllegalArgumentException("Format " + format + " is not supported");
+        }
+    }
+
+    public static CompressionDecoder newDecoder(CompressionFormat format) {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java]: javadocs

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ */
+package io.netty.handler.codec.compression;
+
+/**
+ * Creates a new {@link CompressionEncoder} or a new {@link CompressionDecoder}
+ * with specified {@link CompressionFormat}.
+ */
+public final class CompressionCodecFactory {
+
+    public static CompressionEncoder newEncoder(CompressionFormat format) {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java]: javadocs

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2014 The Netty Project
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: 2015

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionDecoder.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2014 The Netty Project
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionDecoder.java]: 2015

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+     *
+     * @param preferDirect {@code true} if a direct {@link ByteBuf} should be tried to be used as a target for
+     *                     the encoded messages. If {@code false} is used it will allocate a heap
+     *                     {@link ByteBuf}, which is backed by an byte array.
+     */
+    protected CompressionEncoder(CompressionFormat format, boolean preferDirect) {
+        super(preferDirect);
+        if (format == null) {
+            throw new NullPointerException("format");
+        }
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: ObjectUtil.checkNonNull(...)

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        if (!f.isDone()) {
+            // Ensure the channel is closed even if the write operation completes in time.
+            ctx.executor().schedule(new OneTimeTask() {
+                @Override
+                public void run() {
+                    if (!promise.isDone()) {
+                        ctx.close(promise);
+                    }
+                }
+            }, 10, TimeUnit.SECONDS); // FIXME: Magic number
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: what about allow the user to adjust this number via setter ?

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    protected abstract ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise);
+
+    /**
+     * Returns current {@link ChannelHandlerContext}.
+     */
+    protected ChannelHandlerContext ctx() {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: final

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+                }
+            });
+            return p;
+        }
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        ChannelFuture f = finishEncode(ctx, ctx.newPromise());
+        f.addListener(new ChannelFutureListener() {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: only add the listener to the future if its not done. If its done you can just call `ctx.close(...)` directly and so reduce object allocations

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     * The given {@link ChannelFuture} will be notified once the operation
+     * completes and will also be returned.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    public ChannelFuture close(final ChannelPromise promise) {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: final ?

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     *
+     * The returned {@link ChannelFuture} will be notified once the
+     * operation completes.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    public ChannelFuture close() {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: final

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+        }
+    }
+
+    /**
+     * Finishes current encoding.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    protected abstract ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise);
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: Actually the javadoc is not correct as it not throws but just fail the ChannelPromise.

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionFormat.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2014 The Netty Project
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionFormat.java]: 2015

File: codec/src/main/java/io/netty/handler/codec/compression/FastLzFrameEncoder.java [hunk 16]
```diff
@@ -183,4 +185,29 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
...
+    }
+
+    @Override
+    public ChannelFuture close() {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
+    }
+
+    @Override
+    public ChannelFuture close(ChannelPromise promise) {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/FastLzFrameEncoder.java]: `return promise.setFailure(new UnsupportedOperationException())`

File: codec/src/main/java/io/netty/handler/codec/compression/LzfEncoder.java [hunk 17]
```diff
@@ -137,4 +138,29 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
...
+    }
+
+    @Override
+    public ChannelFuture close() {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
+    }
+
+    @Override
+    public ChannelFuture close(ChannelPromise promise) {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/LzfEncoder.java]: `return promise.setFailure(new UnsupportedOperationException())`

File: codec/src/main/java/io/netty/handler/codec/compression/LzfEncoder.java [hunk 18]
```diff
@@ -137,4 +138,29 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
...
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.close(promise);
+    }
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return ctx.newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/LzfEncoder.java]: return promise.setFailure(new UnsupportedOperationException())

File: codec/src/main/java/io/netty/handler/codec/compression/FastLzFrameEncoder.java [hunk 19]
```diff
@@ -183,4 +185,29 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
...
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.close(promise);
+    }
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return ctx.newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/FastLzFrameEncoder.java]: `return promise.setFailure(new UnsupportedOperationException())`

File: codec/src/main/java/io/netty/handler/codec/compression/LzmaFrameEncoder.java [hunk 20]
```diff
@@ -212,4 +214,29 @@ private static int maxOutputBufferLength(int inputLength) {
...
+    }
+
+    @Override
+    public ChannelFuture close() {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
+    }
+
+    @Override
+    public ChannelFuture close(ChannelPromise promise) {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/LzmaFrameEncoder.java]: return promise.setFailure(new UnsupportedOperationException())

File: codec/src/main/java/io/netty/handler/codec/compression/LzmaFrameEncoder.java [hunk 21]
```diff
@@ -212,4 +214,29 @@ private static int maxOutputBufferLength(int inputLength) {
...
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.close(promise);
+    }
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return ctx.newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/LzmaFrameEncoder.java]: return promise.setFailure(new UnsupportedOperationException())

File: codec/src/main/java/io/netty/handler/codec/compression/SnappyFrameEncoder.java [hunk 22]
```diff
@@ -121,4 +126,29 @@ private static void writeChunkLength(ByteBuf out, int chunkLength) {
...
+    }
+
+    @Override
+    public ChannelFuture close() {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
+    }
+
+    @Override
+    public ChannelFuture close(ChannelPromise promise) {
+        return ctx().newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/SnappyFrameEncoder.java]: return promise.setFailure(new UnsupportedOperationException())

File: codec/src/main/java/io/netty/handler/codec/compression/SnappyFrameEncoder.java [hunk 23]
```diff
@@ -121,4 +126,29 @@ private static void writeChunkLength(ByteBuf out, int chunkLength) {
...
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.close(promise);
+    }
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return ctx.newFailedFuture(new UnsupportedOperationException());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/SnappyFrameEncoder.java]: return promise.setFailure(new UnsupportedOperationException())

**Comment by normanmaurer** [comment]
@idelpivnitskiy sorry for the delay. Just left comments.

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java (hunk 12 above)
Comment (idelpivnitskiy) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: no, this method is overridden in some subclasses

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java (hunk 13 above)
Comment (idelpivnitskiy) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: no, this method is overridden in some subclasses

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+/**
+ * Compresses a {@link ByteBuf} using one of compression algorithms.
+ */
+public abstract class CompressionEncoder extends MessageToByteEncoder<ByteBuf> {
+    /**
+     * Default delay time for close. See method {@link #closeDelay(int)}.
+     */
+    public static final int DEFAULT_CLOSE_DELAY = 10;
+
+    private int closeDelay;
```
Comment (idelpivnitskiy) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: Should it be `volatile`?

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    public abstract boolean isClosed();
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     *
+     * The returned {@link ChannelFuture} will be notified once the operation completes.
+     *
+     * Note: for some {@link CompressionFormat}'s which does not support close operation
+     * this method could return failed {@link ChannelFuture}.
```
Comment (idelpivnitskiy) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: Should I add javadoc to all overriden methods, that they return failed ChannelFuture? Like this https://github.com/netty/netty/pull/2527/files#diff-3411b3b6db61cc9b3639d77d4e421e7eR196

File: codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java (hunk 24 above)
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java]: Yes

**Comment by idelpivnitskiy** [comment]
@normanmaurer done

**Comment by Scottmitch** [comment]
@trustin - This is assigned to you. Can you take care (or re-assign)?

**Comment by trustin** [comment]
Let me take a look. Thanks for the heads up, @Scottmitch !

**Comment by idelpivnitskiy** [comment]
@normanmaurer why is it closed? not accepted?

**Comment by normanmaurer** [comment]
I guess because it was against master and we renamed it. Can you reopen against 4.1?

**Comment by idelpivnitskiy** [comment]
@normanmaurer sure, done #4607

**Pull request closed without merging**